File: Experiment-12-VLSM subnetting/vlsm.py
import ipaddress

network = input("Enter Network (e.g. 192.168.10.0/24): ")

n = int(input("Enter Number of Departments: "))

hosts = []

for i in range(n):
    h = int(input(f"Hosts required for Department {i+1}: "))
    hosts.append(h)

hosts.sort(reverse=True)

current_network = ipaddress.ip_network(network, strict=False)

current_ip = int(current_network.network_address)

print("\nVLSM Allocation\n")

for h in hosts:

    needed = h + 2

    bits = 0

    while (2 ** bits) < needed:
        bits += 1

    prefix = 32 - bits

    subnet = ipaddress.ip_network((current_ip, prefix), strict=False)

    print("Hosts Required :", h)
    print("Network :", subnet.network_address)
    print("Broadcast :", subnet.broadcast_address)
    print("Subnet Mask :", subnet.netmask)
    print("Usable Hosts :", subnet.num_addresses - 2)
    print("-" * 40)

    current_ip += subnet.num_addresses

File: Experiment-12-VLSM subnetting/vlsm.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int main()
{
    int n;

    cout<<"Enter Number of Departments: ";
    cin>>n;

    vector<int> hosts(n);

    for(int i=0;i<n;i++)
    {
        cout<<"Hosts Required for Department "<<i+1<<": ";
        cin>>hosts[i];
    }

    sort(hosts.begin(), hosts.end(), greater<int>());

    int a=192,b=168,c=1,current=0;

    cout<<"\nVLSM Allocation\n";

    for(int i=0;i<n;i++)
    {
        int needed = hosts[i] + 2;

        int bits=0;

        while(pow(2,bits)<needed)
            bits++;

        int block = pow(2,bits);

        int prefix = 32 - bits;

        cout<<"\nDepartment "<<i+1<<endl;

        cout<<"Hosts Required : "<<hosts[i]<<endl;

        cout<<"Subnet Mask : /"<<prefix<<endl;

        cout<<"Network Address : "
            <<a<<"."<<b<<"."<<c<<"."<<current<<endl;

        cout<<"Broadcast Address : "
            <<a<<"."<<b<<"."<<c<<"."<<(current+block-1)<<endl;

        cout<<"Host Range : "
            <<a<<"."<<b<<"."<<c<<"."<<(current+1)
            <<" - "
            <<a<<"."<<b<<"."<<c<<"."<<(current+block-2)
            <<endl;

        current += block;
    }

    return 0;
}